src/NServiceBus.Core/Transports/TransportReceiver.cs

```diff
@@ -44,42 +55,154 @@ public Task Start()
 
-            return Task.FromResult(0);
+            return TaskEx.CompletedTask;
         }
 
-        public async Task Stop()
+        public async Task StartRateLimiting()
         {
-            if (!isStarted)
+            if (state == TransportReceiverState.StartedInRateLimitMode)
             {
                 return;
             }
 
-            try
+            if (await semaphoreSlim.WaitAsync(TimeSpan.FromMilliseconds(50)).ConfigureAwait(false))
+            {
+                try
+                {
+                    if (state == TransportReceiverState.StartedInRateLimitMode)
+                    {
+                        return;
+                    }
+                    await StopAndDisposeReceiver().ConfigureAwait(false);
+
+                    receiver = receiverFactory();
+
+                    await receiver.Init(c => pipelineExecutor.Invoke(c), c => recoverabilityExecutor.Invoke(c), criticalError, rateLimitPushSettings).ConfigureAwait(false);
+                    receiver.Start(rateLimitPushRuntimeSettings);
+
+                    state = TransportReceiverState.StartedInRateLimitMode;
+
+                    try
+                    {
+                        await rateLimitModeStartedNotification.Raise(new RateLimitStarted()).ConfigureAwait(false);
+                    }
+                    catch (Exception exception)
+                    {
+                        Logger.WarnFormat("Failed to enter system outage mode: {0}", exception);
+                    }
+                }
+                finally
+                {
+                    semaphoreSlim.Release();
+                }
+            }
+        }
+
+        public async Task<bool> StopRateLimiting()
+        {
+            if (state == TransportReceiverState.StartedRegular)
+            {
+                return true;
+            }
+
+            if (await semaphoreSlim.WaitAsync(TimeSpan.FromMilliseconds(50)).ConfigureAwait(false))
+            {
+                try
+                {
+                    if (state == TransportReceiverState.StartedRegular)
+                    {
+                        return true;
+                    }
+
+                    await StopAndDisposeReceiver().ConfigureAwait(false);
+                    await Init().ConfigureAwait(false);
+                    await Start().ConfigureAwait(false);
+
+                    try
+                    {
+                        await rateLimitModeEndedNotification.Raise(new RateLimitEnded()).ConfigureAwait(false);
+                    }
+                    catch (Exception exception)
+                    {
+                        Logger.WarnFormat("Failed to stop system outage mode: {0}", exception);
+                    }
+
+                    return true;
+                }
+                finally
+                {
+                    semaphoreSlim.Release();
+                }
+            }
+            return false;
+        }
+
+        public async Task Stop()
```

Comment:
I think we shouldn't leak those details into the receiver. Pretty sure the actual transport pumps need to deal with starting and stopping since we have NSB.Raw and are using this trickery in SC as well. So I would suggest to only deal here with potential concurrent start and stop so that they can't mess up things (if even necessary) and as said before pass in the pushruntimesettings from outside.